feat(defaults): enable 'termguicolors' by default when supported by terminal

gpanders opened the pull request:
Enable `'termguicolors'` automatically when Nvim can detect that truecolor is supported by the host terminal.

If `$COLORTERM` is set to "truecolor" or "24bit", or the terminal's terminfo entry contains capabilities for Tc, RGB, or setrgbf and setrgbb, then we assume that the terminal supports truecolor. Otherwise, the terminal is queried (using both XTGETTCAP and SGR + DECRQSS). If the terminal's response to these queries (if any) indicates that it supports truecolor, then `'termguicolors'` is enabled.

---

This follows the process documented in https://github.com/termstandard/colors#checking-for-colorterm.

Experimental/test results:

`'termguicolors'` is correctly set on:

* iTerm2
* Ghostty
* Wezterm
* Kitty
* Alacritty
* tmux

All of the terminal emulators listed above set `$COLORTERM`, so `'termguicolors'` support is assumed from that. If we explicitly unset `$COLORTERM` and force `TERM=xterm-256color` (mimicking an SSH session lacking proper terminfo, for instance), the results are all still correct _except_ for Alacritty, which does not include `Tc`, `RGB`, `setrgbf`, or `setrgbb` in its terminfo and does not support XTGETTCAP or DECRQSS.

`'termguicolors'` is correctly *not* set on:

* Terminal.app

I was not able to directly test foot or Xterm. I'm confident foot will work. Xterm _should_ work since it parses DECRQSS (and I even bent over backwards to ensure we parse Xterm's DECRQSS response, since it differs from everyone else's).

gpanders committed:
feat(defaults): enable 'termguicolors' by default when supported by terminal

Enable 'termguicolors' automatically when Nvim can detect that truecolor
is supported by the host terminal.

If $COLORTERM is set to "truecolor" or "24bit", or the terminal's
terminfo entry contains capabilities for Tc, RGB, or setrgbf and
setrgbb, then we assume that the terminal supports truecolor. Otherwise,
the terminal is queried (using both XTGETTCAP and SGR + DECRQSS). If the
terminal's response to these queries (if any) indicates that it supports
truecolor, then 'termguicolors' is enabled.

echasnovski commented:
I can confirm that [stock `st`](https://st.suckless.org/) does set `termguicolors` by default and it does indeed support it.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -6495,6 +6495,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 	Requires an ISO-8613-3 compatible terminal.
 
+	Nvim will automatically attempt to determine if the host terminal
+	supports 24-bit color and will enable this option if it does (unless
```
```suggestion
	supports 24-bit color and will disable this option if it does not (unless
```

gpanders replied:
Updated to change the hardcoded default to enable.

justinmk commented:
Would it reduce potential "flicker" if we [actually make this](https://github.com/neovim/neovim/issues/21342#issuecomment-1838962241) the hardcoded default? https://github.com/neovim/neovim/blob/1fd29a28841dee3d25ff079eb24fc160eb02cb3c/src/nvim/options.lua#L8762

So the detection logic is used to disable the default, if necessary. But the common case is that 'tgc' is enabled and don't need to redo its side-effects.

justinmk commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -165,91 +165,92 @@ do
-  ---   OSC 11 ; rgba:<red>/<green>/<blue>/<alpha>
-  ---
-  --- where
+if tty then
```
Probably a good time to introduce a `_tui.lua` file, since we will be adding more and more TUI stuff here.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -1752,6 +1757,44 @@ static int unibi_find_ext_bool(unibi_term *ut, const char *name)
+/// 1. If $COLORTERM is "24bit" or "truecolor", return true
+/// 3. Else, check terminfo for Tc, RGB, setrgbf, or setrgbb capabilities. If
+///    found, return true
+/// 4. Else, return false
```
Looks like "2." is missing :) Should we mention `_tui.lua` / `_tty.lua` here?

gpanders replied:
Fixed, thanks

gpanders committed:
test: set notermguicolors in tests

Set 'notermguicolors' in tests which spawn a child Nvim process to force
existing tests to use 16 colors. Also refactor the child process
invocation to make things a little bit less messy.

gpanders commented:
> Would it reduce potential "flicker" if we [actually make this](https://github.com/neovim/neovim/issues/21342#issuecomment-1838962241) the hardcoded default?
> 
> https://github.com/neovim/neovim/blob/1fd29a28841dee3d25ff079eb24fc160eb02cb3c/src/nvim/options.lua#L8762
> 
> So the detection logic is used to disable the default, if necessary. But the common case is that 'tgc' is enabled and don't need to redo its side-effects.

Either way it has no effect on the flicker. But I agree that `'termguicolors'` being enabled will be the common case, so better to optimize for that. I've changed the hardcoded default to enabled now.

>Probably a good time to introduce a _tui.lua or _tty.lua file, since we will be adding more and more TUI stuff here.

Maybe. For now all of this stuff is still just part of "defaults" though.

gpanders commented:
While updating the tests in `tui_spec.lua` I did a little bit of refactoring to remove some of the messy and difficult to understand/modify string munging and eliminated a few old `TODO`s along the way. Hopefully the tui_spec tests are slightly less scary now.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -98,6 +98,9 @@ The following changes may require adaptations in user config or plugins.
   Use `:colorscheme vim` to revert to the old legacy color scheme.
 
+• 'termguicolors' is enabled by default when Nvim is able to determine that
+  the host terminal emulator supports 24-bit color.
```
also vim_diff.txt

gpanders commented:
> > Would it reduce potential "flicker" if we [actually make this](https://github.com/neovim/neovim/issues/21342#issuecomment-1838962241) the hardcoded default?
> > https://github.com/neovim/neovim/blob/1fd29a28841dee3d25ff079eb24fc160eb02cb3c/src/nvim/options.lua#L8762
> > 
> > So the detection logic is used to disable the default, if necessary. But the common case is that 'tgc' is enabled and don't need to redo its side-effects.
> 
> Either way it has no effect on the flicker. But I agree that `'termguicolors'` being enabled will be the common case, so better to optimize for that. I've changed the hardcoded default to enabled now.

Actually, I need to think about this more. This gets tricky in the case that `$COLORTERM` and terminfo aren't available:

1. In that case, we disable `'termguicolors'`
2. We then query the terminal to see if it supports `'termguicolors'`, and enable it if it does, but only if the user hasn't already set `'termguicolors'`
3. But because _we ourselves_ set `'termguicolors'` in (1), we can no longer tell if the user has set it!

justinmk commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -311,4 +312,108 @@ do
+    --- @param enable boolean If true, enable 'termguicolors', else disable
+    local function settgc(enable)
+      if not vim.api.nvim_get_option_info2('termguicolors', {}).was_set then
+        vim.o.termguicolors = enable
```
Looks like `settgc` is only ever called with `true`. Also, it looks possible that `settgc` is called twice, but the 2nd call will be ignored, is that intentional?

gpanders replied:
`settgc` can be called in indeterminate order, depending on when responses from the terminal arrive. We only want to set it if not already set by the user, which is why it's guarded.

But this is broken right now anyway, see my [earlier comment](https://github.com/neovim/neovim/pull/26407#issuecomment-1841743997). I need to think about this some more.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -1752,6 +1757,44 @@ static int unibi_find_ext_bool(unibi_term *ut, const char *name)
+/// 2. Else, check terminfo for Tc, RGB, setrgbf, or setrgbb capabilities. If
+///    found, return true
+/// 3. Else, return false
+static bool term_supports_truecolor(TUIData *tui, const char *colorterm)
```
```suggestion
static bool term_has_truecolor(TUIData *tui, const char *colorterm)
```

justinmk commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -311,4 +312,108 @@ do
     end)
   end
+
+  --- If the TUI was unable to determine that the host terminal supports truecolor, disable
```
```suggestion
  --- If the TUI (term_has_truecolor) was unable to determine that the host terminal supports truecolor, disable
```

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -2208,14 +2234,17 @@ describe("TUI 't_Co' (terminal colors)", function()
+      '--cmd', nvim_set,
+    }, {
+      env = {
+        TERM = term or '',
```
missing LANG=C ?

justinmk commented:
> 3\. But because _we ourselves_ set `'termguicolors'` in (1), we can no longer tell if the user has set it!

But if it's already enabled, why do we need to set it again? And conversely, if the user disabled it, we can skip/ignore all of our detection stuff.

gpanders commented:
> > 3. But because _we ourselves_ set `'termguicolors'` in (1), we can no longer tell if the user has set it!
> 
> But if it's already enabled, why do we need to set it again?

We disable it if we can't tell that the terminal emulator supports truecolor (i.e. when `$COLORTERM` is undefined or not set to `truecolor`/`24bit` and when terminfo doesn't contain `Tc`/`RGB`). We then re-enable it only if querying the terminal shows that the terminal actually does support truecolor. We cannot wait until the query is finished to disable `'tgc'` because that would result in a visible (and annoying) flicker effect on terminals which don't support truecolor.

Keeping the hardcoded default `false` and only _enabling_ if we detect support avoids this problem (this is what I had implemented before [this comment](https://github.com/neovim/neovim/pull/26407#issuecomment-1841729760)). I may revert back to that if I can't think of a solution here.

>And conversely, if the user disabled it, we can skip/ignore all of our detection stuff.

The detection stuff runs before any user config files, so it's not possible for the user to set it before that happens. We have to check if the user has disabled it when the responses to our terminal queries arrive.

gpanders commented:
Reverting the hardcoded default back to `false` and enabling in `_defaults.lua` (instead of vice versa). This works around the issues described above (and the visible effect for the user is identical).

gpanders commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -2540,14 +2572,13 @@ describe("TUI", function()
-      extra_args or "",
-      nvim_set)
-    screen = thelpers.screen_setup(0, cmd)
+    screen = setup_child_nvim({
```
Thanks, fixed.

gpanders commented:
Last failing test:

```
FAILED   1 test, listed below:
FAILED   test/functional/terminal/tui_spec.lua @ 1851: TUI argv[0] can be overridden #23953
test/functional/terminal/tui_spec.lua:1865: Row 1 did not match.
Expected:
  |*{1: }                                                 |
  |*{4:~                                                 }|
  |*{4:~                                                 }|
  |*{4:~                                                 }|
  |*{5:[No Name]                       0,0-1          All}|
  |                                                  |
  |{3:-- TERMINAL --}                                    |
Actual:
  |*                                                  |
  |*[Process exited 0]{1: }                               |
  |*                                                  |
  |*                                                  |
  |*                                                  |
  |                                                  |
  |{3:-- TERMINAL --}                                    |
```

But this fails for me on master as well.

lewis6991 commented:
Great work on the turnaround time for this!

gpanders commented:
@justinmk should be fixed by https://github.com/neovim/neovim/pull/26437
